pkg/controllers/provisioning/v1alpha1/allocator/constraints.go

```diff
@@ -15,50 +15,135 @@ limitations under the License.
 	}
-	return groups
+	return groups, nil
 }
 
-// TODO
-func matchesConstraints(constraints *cloudprovider.Constraints, pod *v1.Pod) bool {
-	return false
+func (c *Constraints) getConstraints(ctx context.Context, spec *v1alpha1.ProvisionerSpec, pod *v1.Pod) (*cloudprovider.Constraints, error) {
+	nodeConstraints := c.getNodeConstraints(spec, pod)
+	// Uses a theoretical node object to verify scheduling properties. This
+	// won't be saved to the KubeAPI, but is used to compute daemonset overhead.
+	overhead, err := c.getNodeOverhead(ctx, pod, &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{Labels: nodeConstraints.Labels},
+		Spec:       v1.NodeSpec{Taints: spec.Allocation.Taints},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("computing node overhead, %w", err)
+	}
+	return &cloudprovider.Constraints{
+		NodeConstraints: nodeConstraints,
+		Overhead:        overhead,
+		Pods:            []*v1.Pod{},
+	}, nil
 }
 
-func constraintsForPod(pod *v1.Pod) *cloudprovider.Constraints {
-	return &cloudprovider.Constraints{
-		Overhead:     calculateOverheadResources(),
-		Architecture: getSystemArchitecture(pod),
-		Topology:     map[cloudprovider.TopologyKey]string{},
-		Pods:         []*v1.Pod{pod},
+func (c *Constraints) getNodeConstraints(spec *v1alpha1.ProvisionerSpec, pod *v1.Pod) cloudprovider.NodeConstraints {
+	return cloudprovider.NodeConstraints{
+		Labels:          c.getLabels(spec, pod),
+		Architecture:    c.getArchitecture(pod),
+		OperatingSystem: c.getOperatingSystem(pod),
 	}
 }
 
-func calculateOverheadResources() v1.ResourceList {
-	//TODO
-	return v1.ResourceList{}
+func (c *Constraints) getLabels(spec *v1alpha1.ProvisionerSpec, pod *v1.Pod) map[string]string {
+	labels := map[string]string{}
+	for _, l := range []map[string]string{
+		spec.Allocation.Labels,
+		pod.Spec.NodeSelector,
+	} {
+		for k, v := range l {
+			labels[k] = v
+		}
+	}
+	return labels
 }
 
-func getSystemArchitecture(pod *v1.Pod) cloudprovider.Architecture {
-	return cloudprovider.ArchitectureLinux386
+func (c *Constraints) getArchitecture(pod *v1.Pod) cloudprovider.Architecture {
+	if architecture, ok := pod.Spec.NodeSelector[OperatingSystemLabelKey]; ok {
+		return cloudprovider.Architecture(architecture)
+	}
+	return cloudprovider.ArchitectureAmd64
+}
+
+func (c *Constraints) getOperatingSystem(pod *v1.Pod) cloudprovider.OperatingSystem {
+	if operatingSystem, ok := pod.Spec.NodeSelector[OperatingSystemLabelKey]; ok {
+		return cloudprovider.OperatingSystem(operatingSystem)
+	}
+	return cloudprovider.OperatingSystemLinux
+}
+
+func (c *Constraints) getNodeOverhead(ctx context.Context, pod *v1.Pod, node *v1.Node) (v1.ResourceList, error) {
+	// 1. Get DaemonSets
+	daemonSetList := &appsv1.DaemonSetList{}
+	if err := c.kubeClient.List(ctx, daemonSetList); err != nil {
+		return nil, fmt.Errorf("listing DaemonSets, %w", err)
+	}
+
+	// 2. filter DaemonSets to include those that will schedule on this node
+	daemonSets := []appsv1.DaemonSet{}
```

Comment:
I think the answer is "no" but just want to throw this out there anyway - should this slice be sorted (or use a Set object; I think there's one in k8s libraries?) Just in case anyone ever compares two equivalent lists of DaemonSets and somehow they're the same but in different order?

Reply:
This is guaranteed to be a set from the API Server